### ISSUE TITLE
Plugin management.

### ISSUE THREAD

**Issue by sunshinejr** [danger/danger-swift#33]
Hey there! I was thinking about plugin management for `danger-swift` and I wanted to share my thoughts with you.

My first idea was to use SPM and `Package.swift`. Then if you wanted to edit the plugin, we would have to generate the project and so on - not that bad, but I'm not sure I'm feeling it. Another configuration file to handle, which works, but still can get confusing.

Then I thought about Marathon. It could work, but then I have troubles running marathon on my computer, it is another dependency and still another config file, Marathonfile. Probably not that hard to learn, but if people see that they need to learn another manager (even if it's really simple), they may just avoid swift version of danger. 

So I kinda combined both options, and figured that we could use inline dependencies in `Dangerfile.swift`:

```swift
import DangerSwiftlint // https://github.com/ashfurrow/danger-swiftlint
```

Then in `danger-swift edit` command we would decide on what we want to use for project creating and so on (I would be probably more for SPM usage, but it would be an implementation detail and we could change it in the meantime).

That's it I believe. Really interested in what you think, @orta.

**Comment by orta** [comment]
Interestingly, that pattern works in both [Marathon](https://github.com/JohnSundell/Marathon#specifying-dependencies-inline) and [Diamond](https://github.com/johnno1962/Diamond#linux). 

IMO, this is 👍🏾 - so to build this, the runner will need:

* To do a pre-parse of the `Dangerfile.swift` for those URLs ^
* Generate a `Package.swift` with those deps in a temp folder
* Build the deps via `swift build`
* _set up the library paths when compiling the Dangerfile.swift to include the new dependencies_

A tricky aspect of this is that those Danger plugins will need a reference to the `Danger` lib, but that might be able to "just work" via a dep on Danger inside the generated SPM package.

**Comment by orta** [comment]
oh dang, yeah, this totally affects the edit command too - hrrrm

**Comment by orta** [comment]
OK, then I need to re-think the edit command in lieu of ^

**Comment by orta** [comment]
OK, I think that the edit command would rely entirely on the generated `package.swift` - which means I can close this PR. 

There'll need to be a copy of the Dangerfile.swift into that temp folder, maybe we can add a build command to copy the changes back to the original on build ( or use a sym/hard link) 

@JohnSundell - I think you do this with your `edit` command, got any advice if you had to recreate from scratch?

**Comment by JohnSundell** [comment]
This is an awesome idea! 🚀 I love the idea of plugins for everything 😄

One solution could be to have `danger-swift edit` just invoke a Marathon editing session (this is a pattern I'm taking in the upcoming version of SwiftPlate, which will also support plugins, powered by Marathon). The advantage of that would be that you get to utilize all dependency management, etc that Marathon already does, rather than having to re-implement it. Can go more into detail of what I have in mind if you like?

Marathon's editing mode works by copying the original file into the internal Marathon folder, then continuously copying the file back. I wanted to use symlinks (I do that for packages), but Xcode doesn't like to open symlinks I'm afraid 😢. You can find the copy-loop implementation here: https://github.com/JohnSundell/Marathon/blob/master/Sources/MarathonCore/Script.swift#L188

Btw, what problems are you facing when running Marathon on your machine @sunshinejr? 🤔 You also don't need to use a `Marathonfile` with Marathon (it's just one way of declaring dependencies), like @orta mentions, you can do it inline using a comment syntax like the one you're proposing for this 🙂

**Comment by orta** [comment]
Yeah, that's perfect, so long as danger-swift can trigger the dep generation and hook up all the libs correctly when [evaluating the Dangerfile](https://github.com/danger/danger-swift/blob/master/Sources/Runner/main.swift#L72-L99) then that works for me for both edit and running.

**Comment by JohnSundell** [comment]
Yeah 👍 To avoid the additional learning curve that @sunshinejr mentions I'd also suggest making the dependency opaque, and use Marathon as a framework rather than an additional command line tool that the user needs to install.

**Comment by sunshinejr** [comment]
@JohnSundell yeah, making the Marathon dependency implicit would be great 👍 I knew about the inline dependencies from Marathon, actually, that’s why I said I mixed SPM with Marathon :D I think this is really neat idea and I would love to use it. 

About the issue with Marathon I have, I will create an issue on your repo so we don’t spam this one :D

**Comment by orta** [comment]
Had a chat with @ashfurrow - gonna:

- Use Marathon to do inline deps
- Let those deps get built out in their own folder before evaluation
- Use the library paths to connect the Dangerfile.swift compilation to the runtime libs

Sidenote: I wonder if this will completely break the current Danger execution model, to the point where danger-swift will need to actually compile and then execute the Dangerfile, rather than the current swiftc funkiness.

**Comment by orta** [comment]
https://github.com/danger/danger-swift/pull/37 does this, it supports:

```swift
import DangerSwiftlint // package: https://github.com/ashfurrow/danger-swiftlint
```

**Comment by orta** [comment]
Shipped in 0.3.0

**Comment by orta** [comment]
Refined to "working" in 0.3.1
